dali/operators/segmentation/utils/searchable_rle_mask.h

```diff
@@ -0,0 +1,111 @@
+// Copyright (c) 2020, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_OPERATORS_SEGMENTATION_UTILS_SEARCHABLE_RLE_MASK_H_
+#define DALI_OPERATORS_SEGMENTATION_UTILS_SEARCHABLE_RLE_MASK_H_
+
+#include <vector>
+#include "dali/core/tensor_view.h"
+#include "dali/core/span.h"
+
+namespace dali {
+
+/**
+ * @brief Encodes a segmentation mask in a way that is easy to access to
+ *        know how many foreground pixels are there and get the flat index
+ *        to an arbitrary i-th foreground pixel, as they appear in the mask.
+ */
+class SearchableRLEMask {
+ public:
+  struct Group {
+    int64_t ith;    // The group begins with the i-th pixel
+    int64_t start;  // The group starts at this flat position
+  };
+
+  /**
+   * @brief Any value above threshold is consider foreground
+   */
+  template <typename T>
+  explicit SearchableRLEMask(span<const T> mask_view, T threshold = T(0)) {
+    int64_t idx = 0;
+    int64_t sz = mask_view.size();
+    while (idx < sz) {
+      while (idx < sz && mask_view[idx] <= threshold) {
+        idx++;
+      }
+      if (idx < sz) {  // found a foreground pixel
+        groups_.push_back({count_++, idx++});
+        while (idx < sz && mask_view[idx] > threshold) {
+          idx++;
+          count_++;
+        }
+      }
+    }
+  }
+
+  template <typename T>
+  explicit SearchableRLEMask(TensorView<StorageCPU, T> mask_view, T threshold = T(0))
+    : SearchableRLEMask(span<const T>{mask_view.data, volume(mask_view.shape)}, threshold) {}
+
+  /**
+   * @brief Returns the position of the i-th foreground pixel.
+   *        If ith is an invalid index, -1 is returned
+   */
+  int64_t find(int64_t ith) {
+    if (ith < 0 || ith >= count_) {
+      return -1;
+    }
+
+    int64_t start = 0;
+    int64_t end = groups_.size();
+    int64_t last_idx = groups_.size() - 1;
+    while (end != start) {
+      int64_t pos = (end + start) / 2;
```

Comment:
```suggestion
      int64_t pos = (end + start) >> 1;
```
In the inner loop, performance matters.